### ISSUE TITLE
Sudden Gradient Exploding During VQNSP Codebook Pre-training

### ISSUE THREAD

**Issue by alskdjfasdfsadf** [935963004/LaBraM#65]
Hi, Thank you for this outstanding work!
I'm implementing the LaBraM pipeline and have encountered an issue during VQNSP codebook pre-training that I'd appreciate your insights on.

**Setup**
I'm following your methodology exactly but with hardware constraints:
Hardware: 2 GPUs (instead of the recommended 8)
Dataset: TUAB with identical preprocessing to your pipeline
Stage: VQNSP codebook pre-training

**Data Processing**
I've preprocessed the TUH Abnormal EEG Corpus v3.0.1 following your specifications:
Standard 21-channel montage selection
0.1-75Hz bandpass filtering, 50Hz notch filtering
200Hz resampling
10-second segments
Output: tuab_pretrain.hdf5 format

**Training Configuration**
torchrun --nnodes=1 --nproc_per_node=2 run_vqnsp_training.py \
    --model vqnsp_encoder_base_decoder_3x200x12 \
    --codebook_n_emd 8192 \
    --codebook_emd_dim 64 \
    --quantize_kmeans_init \
    --batch_size 256 \
    --opt adamw \
    --opt_betas 0.9 0.99 \
    --weight_decay 1e-4 \
    --warmup_epochs 10 \
    --epochs 100
Effective batch size: 512 (vs. your original 1024)

**Issue Description**
Training proceeded normally for approximately 6000 steps with steady loss reduction (2.02 → 1.20) and stable gradient norms (~0.76). However, at step 6030, I encountered sudden gradient exploding:
Warning: NaN or Inf found in input tensor.
grad_norm: 0.5547 (inf)
The transition was abrupt - no gradual degradation, just sudden instability.

**Questions**
I'm curious about several aspects of this behavior:
Is this a known characteristic of VQNSP training? Do you typically see gradient instability at certain training phases?
Could this be related to the vector quantization mechanism itself? I'm wondering if there's interaction between the EMA updates and gradient flow that becomes problematic as the codebook evolves.
How sensitive is VQNSP training to batch size? Given my reduced effective batch size (512 vs 1024), could this contribute to the instability?
Do you recommend gradient clipping for VQNSP training? I notice it's not specified in your example commands.
Is there a typical "danger zone" during codebook learning where gradients tend to become unstable?

**Technical Context**
Training was in warmup phase (epoch 0 of 10 warmup epochs)
Learning rate at failure: 0.000015
Memory usage remained stable throughout
All loss components were improving normally before the failure

I'm particularly intrigued by the sudden nature of the failure after such stable training. Any insights you could share about the underlying dynamics of VQNSP training would be incredibly helpful.
Thank you for your time and for making this excellent research available!

**Comment by 935963004** [comment]
Hi, this is a common warning. If the loss is continuously decreased, just ignore this warning as it does not affect the training.
